File: apps/backend-java/src/main/java/com/housingplatform/features/admin/mapper/AdminMapper.java
package com.housingplatform.features.admin.mapper;

import com.housingplatform.features.admin.dto.AdminHost;
import com.housingplatform.features.admin.dto.AdminPayment;
import com.housingplatform.features.admin.dto.AdminProperty;
import com.housingplatform.features.admin.dto.AuditLogDto;
import com.housingplatform.features.admin.dto.HousingRequestDto;
import com.housingplatform.persistence.repository.AdminRepositoryCustom.AdminHostRow;
import com.housingplatform.persistence.repository.AdminRepositoryCustom.AdminPaymentRow;
import com.housingplatform.persistence.repository.AdminRepositoryCustom.AdminPropertyRow;
import com.housingplatform.persistence.repository.AdminRepositoryCustom.AuditLogRow;
import com.housingplatform.persistence.repository.AdminRepositoryCustom.HousingRequestRow;
import com.housingplatform.features.hosts.mapper.HostMapper;
import com.housingplatform.persistence.entity.Host;
import com.housingplatform.persistence.entity.HousingRequest;
import com.housingplatform.persistence.enums.AccommodationType;
import com.housingplatform.persistence.enums.HousingRequestStatus;
import com.housingplatform.persistence.enums.PaymentStatus;
import com.housingplatform.persistence.enums.PropertyStatus;
import com.housingplatform.features.properties.dto.PropertyStatusChange;

public final class AdminMapper {

  private AdminMapper() {}

  public static AdminProperty toAdminProperty(AdminPropertyRow row) {
    var base = HostMapper.toPropertyListItem(
        new com.housingplatform.persistence.repository.HostRepositoryCustom.HostPropertyListRow(
            row.id(),
            row.title(),
            row.slug(),
            row.propertyType(),
            row.district(),
            row.status(),
            row.bookingMode(),
            row.monthlyPriceMin(),
            row.roomCount(),
            row.updatedAt()));
    return new AdminProperty(
        base.id(),
        base.title(),
        base.slug(),
        base.propertyType(),
        base.district(),
        base.status(),
        base.bookingMode(),
        base.monthlyPriceMin(),
        base.roomCount(),
        base.updatedAt(),
        row.hostDisplayName());
  }

  public static AdminHost toAdminHost(AdminHostRow row) {
    return new AdminHost(
        row.id(),
        row.displayName(),
        row.status(),
        row.profileName(),
        row.verifiedAt(),
        row.createdAt());
  }

  public static AdminPayment toAdminPayment(AdminPaymentRow row) {
    return new AdminPayment(
        row.id(),
        row.orderId(),
        row.bookingId(),
        row.amountKrw(),
        PaymentStatus.valueOf(row.status()),
        row.propertyTitle(),
        row.customerName(),
        row.confirmedAt(),
        row.createdAt());
  }

  public static HousingRequestDto toHousingRequest(HousingRequestRow row) {
    return new HousingRequestDto(
        row.id(),
        row.email(),
        row.desiredArea(),
        row.checkIn(),
        row.checkOut(),
        row.budgetMax(),
        row.accommodationType() == null
            ? null
            : AccommodationType.fromDbValue(row.accommodationType()),
        row.notes(),
        HousingRequestStatus.fromDbValue(row.status()),
        row.createdAt());
  }

  public static HousingRequestDto toHousingRequest(HousingRequest request) {
    return new HousingRequestDto(
        request.getId(),
        request.getEmail(),
        request.getDesiredArea(),
        request.getCheckIn(),
        request.getCheckOut(),
        request.getBudgetMax(),
        request.getAccommodationType(),
        request.getNotes(),
        request.getStatus(),
        request.getCreatedAt());
  }

  public static AuditLogDto toAuditLog(AuditLogRow row) {
    return new AuditLogDto(
        row.id(),
        row.action(),
        row.entityType(),
        row.entityId(),
        row.actorName(),
        row.metadata(),
        row.createdAt());
  }

  public static PropertyStatusChange toPropertyStatusChange(
      java.util.UUID propertyId, PropertyStatus status) {
    return new PropertyStatusChange(propertyId, status);
  }
}

File: apps/backend-node/src/features/admin/mappers/admin.mapper.ts
import { formatCalendarDate, formatDateTime } from '../../bookings/mappers/booking.mapper';
import { mapHostPropertyListItem } from '../../hosts/mappers/host.mapper';
import { PaymentStatus } from '../../payments/dto/payment-status.dto';
import { propertyStatusToApi } from '../../properties/mappers/property.mapper';
import type { PropertyStatusChange } from '../../properties/dto/property-status-change';
import type { AdminHostDto } from '../dto/admin-host.dto';
import type { AdminPaymentDto } from '../dto/admin-payment.dto';
import type { AdminPropertyDto } from '../dto/admin-property.dto';
import type { AuditLogDto } from '../dto/audit-log.dto';
import type { HousingRequestDto } from '../dto/housing-request.dto';
import { HousingRequestStatus } from '../dto/housing-request-status.dto';

export function mapAdminProperty(row: {
  id: string;
  title: string;
  slug: string;
  propertyType: string;
  district: string;
  status: string;
  bookingMode: string;
  monthlyPriceMin: number | null;
  roomCount: number;
  updatedAt: Date;
  hostDisplayName: string;
}): AdminPropertyDto {
  return {
    ...mapHostPropertyListItem(row),
    hostDisplayName: row.hostDisplayName,
  };
}

export function mapAdminHost(row: {
  id: string;
  displayName: string;
  status: string;
  profileName: string;
  verifiedAt: Date | null;
  createdAt: Date;
}): AdminHostDto {
  return {
    id: row.id,
    displayName: row.displayName,
    status: row.status,
    profileName: row.profileName,
    verifiedAt: row.verifiedAt ? formatDateTime(row.verifiedAt) : null,
    createdAt: formatDateTime(row.createdAt),
  };
}

export function mapAdminPayment(row: {
  id: string;
  orderId: string;
  bookingId: string;
  amountKrw: number;
  status: string;
  propertyTitle: string | null;
  customerName: string | null;
  confirmedAt: Date | null;
  createdAt: Date;
}): AdminPaymentDto {
  return {
    id: row.id,
    orderId: row.orderId,
    bookingId: row.bookingId,
    amountKrw: row.amountKrw,
    status: row.status as PaymentStatus,
    propertyTitle: row.propertyTitle,
    customerName: row.customerName,
    confirmedAt: row.confirmedAt ? formatDateTime(row.confirmedAt) : null,
    createdAt: formatDateTime(row.createdAt),
  };
}

export function mapHousingRequest(row: {
  id: string;
  email: string;
  desiredArea: string;
  checkIn: Date | null;
  checkOut: Date | null;
  budgetMax: number | null;
  accommodationType: string | null;
  notes: string | null;
  status: string;
  createdAt: Date;
}): HousingRequestDto {
  return {
    id: row.id,
    email: row.email,
    desiredArea: row.desiredArea,
    checkIn: row.checkIn ? formatCalendarDate(row.checkIn) : null,
    checkOut: row.checkOut ? formatCalendarDate(row.checkOut) : null,
    budgetMax: row.budgetMax,
    accommodationType: row.accommodationType,
    notes: row.notes,
    status: row.status as HousingRequestStatus,
    createdAt: formatDateTime(row.createdAt),
  };
}

export function mapAuditLog(row: {
  id: string;
  action: string;
  entityType: string;
  entityId: string | null;
  actorName: string;
  metadata: Record<string, unknown>;
  createdAt: Date;
}): AuditLogDto {
  return {
    id: row.id,
    action: row.action,
    entityType: row.entityType,
    entityId: row.entityId,
    actorName: row.actorName,
    metadata: row.metadata,
    createdAt: formatDateTime(row.createdAt),
  };
}

export function mapPropertyStatusChange(
  propertyId: string,
  status: string,
): PropertyStatusChange {
  return {
    id: propertyId,
    status: propertyStatusToApi(status as Parameters<typeof propertyStatusToApi>[0]),
  };
}
